### PR TITLE
bluetooth: services: ancs_client: Use callback functions to report results in GATT operations

### PR DESCRIPTION
**Pull request by leewkb4567**
Use callback functions to report results in GATT operations. The current ANCS client uses an event handler, which is ported from nRF5 SDK.

### CODE REVIEW

**Comment by leewkb4567** [comment]
The ANCS client documentation has usage updated to reflect the code changes.

File: subsys/bluetooth/services/ancs_client.c [hunk 1]
```diff
@@ -368,15 +364,13 @@ static void bt_ancs_cp_write_callback(struct bt_conn *conn, uint8_t err,
...
-
-	ancs_c->evt_handler(&ancs_evt);
+	if (ancs_c->cp_write_cb) {
+		ancs_c->cp_write_cb(ancs_c, err);
+	}
 }
 
-int bt_ancs_cp_write(struct bt_ancs_client *ancs_c, uint16_t len)
+int bt_ancs_cp_write(struct bt_ancs_client *ancs_c, uint16_t len,
+		     bt_ancs_write_cb func)
```
Comment (joerchan) [inline subsys/bluetooth/services/ancs_client.c:373]: Not validating the func not NULL here, is this optional?

Comment (joerchan) [inline subsys/bluetooth/services/ancs_client.c:373]: Also the function is public, but could potentially conflict with ANCS_CP_WRITE_PENDING in progress also using the write parameters, right?

**Comment by joerchan** [comment]
@leewkb4567 Could you also explain why you are doing the change in the commit messages.

File: subsys/bluetooth/services/ancs_client.c (hunk 1 above)
Comment (leewkb4567) [reply subsys/bluetooth/services/ancs_client.c:373]: It is possible that the callback function is NULL. e.g. the user does not check the GATT write response.

A NULL pointer check is done in bt_ancs_cp_write_callback().

A careless mistake is fixed in bt_ancs_cp_write_callback(). The callback function variable should be copied before clearing the atomic bit. It avoids a racing condition in a multi-threading system.

Comment (leewkb4567) [reply subsys/bluetooth/services/ancs_client.c:373]: bt_ancs_cp_write() is a routine to minimize repeated code.

There are several calls to this function. Each sets the ANCS_CP_WRITE_PENDING bit to claim usage of GATT write parameters and a common data buffer. When the GATT write is not done, ANCS_CP_WRITE_PENDING is set and further calls to set the bit gets a -EBUSY error. The function prototype gets a comment to document this.

**Comment by leewkb4567** [comment]
> @leewkb4567 Could you also explain why you are doing the change in the commit messages.

Done. Rebased to current master branch too.

File: subsys/bluetooth/services/ancs_client.c [hunk 2]
```diff
@@ -60,60 +60,53 @@ static int bt_ancs_verify_notification_format(const struct bt_ancs_evt_notif *no
...
 {
-	struct bt_ancs_evt ancs_evt;
 	int err;
+	struct bt_ancs_evt_notif notif;
 
 	if (len != BT_ANCS_NOTIF_DATA_LENGTH) {
-		ancs_evt.evt_type = BT_ANCS_EVT_INVALID_NOTIF;
-		ancs_c->evt_handler(&ancs_evt);
-
+		ancs_c->ns_notif_cb(ancs_c, -EINVAL, &notif);
```
[the remark this answers is not in the document]
Comment (leewkb4567) [reply subsys/bluetooth/services/ancs_client.c]: OK. The NULL pointer check is done by an inline function bt_ancs_do_ns_notif_cb() instead.

File: subsys/bluetooth/services/ancs_client.c [hunk 3]
```diff
@@ -60,60 +60,53 @@ static int bt_ancs_verify_notification_format(const struct bt_ancs_evt_notif *no
...
-	err = bt_ancs_verify_notification_format(&ancs_evt.notif);
-	if (!err) {
-		ancs_evt.evt_type = BT_ANCS_EVT_NOTIF;
-	} else {
-		ancs_evt.evt_type = BT_ANCS_EVT_INVALID_NOTIF;
-	}
+	err = bt_ancs_verify_notification_format(&notif);
 
-	ancs_c->evt_handler(&ancs_evt);
+	ancs_c->ns_notif_cb(ancs_c, err, &notif);
```
[the remark this answers is not in the document]
Comment (leewkb4567) [reply subsys/bluetooth/services/ancs_client.c]: Done by an inline function bt_ancs_do_ns_notif_cb().

File: include/bluetooth/services/ancs_client.rst [hunk 4]
```diff
@@ -37,32 +37,32 @@ Once a connection is established with a central device, the module needs a servi
 If this succeeds, the handles for the ANCS server must be assigned to an ANCS client instance using the :c:func:`bt_ancs_handles_assign` function.
 
 The application can now subscribe to iOS notifications with :c:func:`bt_ancs_subscribe_notification_source`.
-The notifications arrive in the :c:enumerator:`BT_ANCS_EVT_NOTIF` event.
+The notifications arrive through callback function of the notification source subscription.
```
Comment (peknis) [inline include/bluetooth/services/ancs_client.rst:40]:
```suggestion
The notifications arrive through the callback function of the notification source subscription.
```

File: include/bluetooth/services/ancs_client.rst [hunk 5]
```diff
@@ -37,32 +37,32 @@ Once a connection is established with a central device, the module needs a servi
...
 Use :c:func:`bt_ancs_request_attrs` to request attributes for the notifications.
-The attributes arrive in the :c:enumerator:`BT_ANCS_EVT_NOTIF_ATTRIBUTE` event.
+The attributes arrive through callback function of the data source subscription with a Command ID :c:enumerator:`BT_ANCS_COMMAND_ID_GET_NOTIF_ATTRIBUTES`.
```
Comment (peknis) [inline include/bluetooth/services/ancs_client.rst:42]:
```suggestion
The attributes arrive through the callback function of the data source subscription with a Command ID :c:enumerator:`BT_ANCS_COMMAND_ID_GET_NOTIF_ATTRIBUTES`.
```

File: include/bluetooth/services/ancs_client.rst [hunk 6]
```diff
@@ -37,32 +37,32 @@ Once a connection is established with a central device, the module needs a servi
...
 Use :c:func:`bt_ancs_request_app_attr` to request attributes of the app that issued the notifications.
```
Comment (peknis) [inline include/bluetooth/services/ancs_client.rst:43]:
```suggestion
Use :c:func:`bt_ancs_request_app_attr` to request attributes of the application that issued the notifications.
```

File: include/bluetooth/services/ancs_client.rst [hunk 7]
```diff
@@ -37,32 +37,32 @@ Once a connection is established with a central device, the module needs a servi
...
-The app attributes arrive in the :c:enumerator:`BT_ANCS_EVT_APP_ATTRIBUTE` event.
+The app attributes arrive through callback function of the data source subscription with a Command ID :c:enumerator:`BT_ANCS_COMMAND_ID_GET_APP_ATTRIBUTES`.
```
Comment (peknis) [inline include/bluetooth/services/ancs_client.rst:44]:
```suggestion
The application attributes arrive through the callback function of the data source subscription with a Command ID :c:enumerator:`BT_ANCS_COMMAND_ID_GET_APP_ATTRIBUTES`.
```

**Review by peknis — CHANGES_REQUESTED** [review]

File: include/bluetooth/services/ancs_client.rst (hunk 4 above)
Comment (leewkb4567) [reply include/bluetooth/services/ancs_client.rst]: Done

File: include/bluetooth/services/ancs_client.rst (hunk 5 above)
Comment (leewkb4567) [reply include/bluetooth/services/ancs_client.rst]: Done

File: include/bluetooth/services/ancs_client.rst (hunk 6 above)
Comment (leewkb4567) [reply include/bluetooth/services/ancs_client.rst:43]: The app refers to an iPhone app. Let's stick to the Apple terminology.

File: include/bluetooth/services/ancs_client.rst (hunk 7 above)
Comment (leewkb4567) [reply include/bluetooth/services/ancs_client.rst:44]: The app refers to an iPhone app.

File: include/bluetooth/services/ancs_client.rst (hunk 6 above)
Comment (peknis) [reply include/bluetooth/services/ancs_client.rst:43]: Fair enough.

**Comment by leewkb4567** [comment]
Rebase to current master.

**Review by peknis — APPROVED** [review]

File: include/bluetooth/services/ancs_client.rst (hunk 7 above)
Comment (leewkb4567) [reply include/bluetooth/services/ancs_client.rst]: Overlooked "the callback function". Fixed now.

**Review by joerchan — APPROVED** [review]

**Review by KAGA164 — APPROVED** [review]

**Pull request merged**